File: solutions/python/queen-attack/1/queen_attack.py
class Queen:
    def __init__(self, row, column):
        if row < 0:
            raise ValueError("row not positive")
        if row > 7:
            raise ValueError("row not on board")
        if column < 0:
            raise ValueError("column not positive")
        if column > 7:
            raise ValueError("column not on board")
            
        self.row = row
        self.column = column

    def can_attack(self, another_queen):
        # Check if queens are in the same position
        if self.row == another_queen.row and self.column == another_queen.column:
            raise ValueError("Invalid queen position: both queens in the same square")
            
        # Same row or column
        if self.row == another_queen.row or self.column == another_queen.column:
            return True
            
        # Diagonal attack - check if the absolute difference between rows equals 
        # the absolute difference between columns
        row_diff = abs(self.row - another_queen.row)
        col_diff = abs(self.column - another_queen.column)
        return row_diff == col_diff

File: solutions/dart/queen-attack/1/lib/queen_attack.dart
class Queen {
  final int row;
  final int col;

  Queen(this.row, this.col) {
    // Validate the position of the queen
    assert(row >= 0, 'row not positive');
    assert(row <= 7, 'row not on board');
    assert(col >= 0, 'column not positive');
    assert(col <= 7, 'column not on board');
  }

  bool canAttack(Queen other) {
    // Check if the other piece is in the same row, column, or either diagonal
    return row == other.row ||
        col == other.col ||
        (row - other.row).abs() ==
            (col - other.col).abs() || // Diagonal check (positive slope)
        (row + col) ==
            (other.row + other.col); // Diagonal check (negative slope)
  }
}
